edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs

```diff
@@ -0,0 +1,255 @@
+            if (keySegments.Count() == 2)
+            {
+                return SanitizeDNSDomain(keySegments[0]) + "/" + SanitizeNameValue(keySegments[1]);
+            }
+            else
+            {
+                return SanitizeNameValue(key);
+            }
+        }
+
+        // Alphanumeric, '-' and '.' up to 253 characters.
+        public static string SanitizeK8sValue(string name)
+        {
+            if (name == null)
+            {
+                // Values sometimes may be null, and that's OK.
+                return name;
+            }
+
+            name = name.ToLower();
+
+            var output = new StringBuilder();
+            for (int i = 0; i < name.Length; i++)
+            {
+                if (AllowedGenericHashSet.Contains(name[i]))
+                {
+                    output.Append(name[i]);
+                }
+            }
+
+            if (output.Length > MaxK8SValueLength)
+            {
+                throw new InvalidKubernetesNameException($"Value '{name}' as sanitized exceeded maximum length {MaxK8SValueLength}");
+            }
+
+            return output.ToString();
+        }
+
+        // DNS label (as per RFC 1035)
+        public static string SanitizeDNSValue(string name)
+        {
+            // The name returned from here must conform to following rules (as per RFC 1035):
+            //  - length must be <= 63 characters
+            //  - must be all lower case alphanumeric characters or '-'
+            //  - must start with an alphabet
+            //  - must end with an alphanumeric character
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new InvalidKubernetesNameException($"Name '{name}' is null or empty");
+            }
+
+            name = name.ToLower();
+
+            // get index of first character from the left that is an alphabet
+            int start = 0;
+            while (start < name.Length && !IsAlpha(name[start]))
+            {
+                start++;
+            }
+
+            // get index of last character from right that's an alphanumeric
+            int end = Math.Max(start, name.Length - 1);
+            while (end > start && !IsAlphaNumeric(name[end]))
+            {
+                end--;
+            }
+
+            // build a new string from start-end (inclusive) excluding characters
+            // that aren't alphanumeric or the symbol '-'
+            var output = new StringBuilder();
+            for (int i = start; i <= end; i++)
+            {
+                if (AllowedDnsHashSet.Contains(name[i]))
+                {
+                    output.Append(name[i]);
+                }
+            }
+            if (output.Length > MaxDnsNameLength)
+            {
+                throw new InvalidKubernetesNameException($"DNS name '{name}' exceeded maximum length of {MaxDnsNameLength}");
+            }
+
+            return output.ToString();
```

Comment:
Is it ok to have empty string after sanitization? I see the same pattern in each public function. Could you double check each of them, please?

Reply:
This may need to be tweaked as we need, but if it's throwing on null at the top, then it should also check for empty strings at the end.